Cubical/Categories/Constructions/Hyperdoctrine.agda

```diff
@@ -0,0 +1,63 @@
+{-# OPTIONS --safe #-}
+
+module Cubical.Categories.Constructions.Hyperdoctrine where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Categories.Category
+open import Cubical.Categories.Functor.Base
+open import Cubical.Categories.NaturalTransformation.Base
+
+open import Cubical.Categories.Instances.PosetCat.Base
+open import Cubical.Categories.Instances.PosetCat.Monotone
+open import Cubical.Categories.Instances.PosetCat.MonotoneAdjoint
+open import UMP
+
+private
+  variable
+    ℓ ℓ' : Level
+
+open CartesianCategory
+open Category
+open MonFunAdj
+
+record Hyperdoctrine ℓ ℓ' : Type (ℓ-suc (ℓ-max ℓ ℓ')) where
+  field
+    cc : CartesianCategory ℓ ℓ'
+    func : Functor ((cc .cat) ^op) (PosetCatAdjCycl {ℓ} {ℓ'})
+    {-
+    Beck-Chevalley condition
+    Expresses the naturality of the left/right adjoints of the a × _ projection:
+    For f ∈ cc [b , c], we have the following commuting square:
+
+                              func(π₂)
+                 func(a × c) ⟵--------- func(c)
+                   |         ---------→   |
+      func(1 × f)  |         left-adj     | func(f)
+                   |                      |
+                   ↓          func(π₂)    ↓
+                 func(a × b) ←--------- func(b)
+                             ---------→
+                             left-adj
+
+    TODO: Can probably be more cleanly expressed as a bifunctor
+    -}
+    nat-left : ∀ {b c : cc .cat .ob}
+      (a : cc .cat .ob) (f : cc .cat [ b , c ]) →
+      ((MonComp
+        ((func ⟪ (prod₂-I cc a c {bin-prod-ob cc a b} (π₁ cc a b) ((π₂ cc a b) ⋆⟨ cc .cat ⟩ f)) ⟫) .morphism)
```

Comment:
These definitions should look cleaner using `BinProduct.More.Notation`

Reply:
Finished the refactor with Limits.BinProduct